engine/src/main/java/org/hibernate/validator/internal/engine/tracking/PredefinedScopeProcessedBeansTrackingStrategy.java

```diff
@@ -0,0 +1,368 @@
+					// beanClassesInPath should have already been cleared.
+					assert beanClassesInPath.isEmpty();
+					return true;
+				}
+				// directCascadedBeanClass does not have a cycle.
+				// directCascadedBeanClass would have already been removed by the
+				// call to #determineTrackingRequired above
+			}
+			beanClassesInPath.remove( beanClass );
+			return register( beanClass, false );
+		}
+
+		// TODO: is there a more concise way to do this?
+		private <T> Set<Class<?>> getDirectCascadedBeanClasses(Class<T> beanClass) {
+			final Set<Class<?>> directCascadedBeanClasses = new HashSet<>();
+			// At runtime, if we are not looking at the root bean the actual value of a cascadable
+			//  can be either the same `beanClass` or one of its subtypes... since subtypes can potentially add
+			//  more constraints we want to iterate through the subclasses (for which there is some metadata defined)
+			//  and include the info from them too.
+			Set<Class<?>> classes = subtypesMap.get( beanClass );
+			if ( classes == null ) {
+				// It may be that some bean property without any constraints is marked for cascading validation,
+				//  In that case the metadata entry will be missing from the map:
+				return Set.of();
+			}
+			for ( Class<?> otherBeanClass : classes ) {
+				final BeanMetaData<?> beanMetaData = rawBeanMetaDataMap.get( otherBeanClass );
+
+				if ( beanMetaData == null || !beanMetaData.hasCascadables() ) {
+					continue;
+				}
+
+				for ( Cascadable cascadable : beanMetaData.getCascadables() ) {
+					processSingleCascadable( cascadable, directCascadedBeanClasses );
+				}
+			}
+			return directCascadedBeanClasses;
+		}
+
+		private boolean register(Class<?> beanClass, boolean isBeanTrackingEnabled) {
+			if ( classToBeanTrackingEnabled.put( beanClass, isBeanTrackingEnabled ) != null ) {
+				throw new IllegalStateException( beanClass.getName() + " registered more than once." );
+			}
+			return isBeanTrackingEnabled;
+		}
+	}
+
+	private static void processSingleCascadable(Cascadable cascadable, Set<Class<?>> directCascadedBeanClasses) {
+		CascadingMetaData cascadingMetaData = cascadable.getCascadingMetaData();
+		if ( cascadingMetaData.isContainer() ) {
+			final ContainerCascadingMetaData containerCascadingMetaData = cascadingMetaData.as( ContainerCascadingMetaData.class );
+			processContainerCascadingMetaData( containerCascadingMetaData, directCascadedBeanClasses );
+		}
+		else if ( cascadingMetaData instanceof PotentiallyContainerCascadingMetaData potentiallyContainerCascadingMetaData ) {
+			// if it's a potentially container cascading one, we are "in trouble" as thing can be "almost anything".
+			// TODO: would it be enough to just take the type as defined ?
+			//  directCascadedBeanClasses.add( (Class<?>) cascadable.getCascadableType() );
+			//
+			// TODO: or be much more cautious and just assume that it can be "anything":
+			directCascadedBeanClasses.add( Object.class );
```

Comment:
My thoughts here are that if we are unsure if the tracking could be required or not we should stay on the safe side and assume it would be... 
these potentially-container-cascading cases are trouble anyway so I'd hope users would stay away from them ... 

btw, this also reminds me of the other "problem", once we drop the support for `@Valid List<MyBean>`, things with these potentially cascading cases won't really work. I've started adding some thoughts on that here: https://github.com/jakartaee/validation/issues/266